### PR TITLE
Support Request, Response and User for minimal actions

### LINKED ISSUE #33870: Support ClaimsPrincipal as a parameter with minimal APIs?
### Is your feature request related to a problem? Please describe.

I've been playing around with the new minimal APIs in a toy TodoApp sample application where for demonstration purposes the Todo items belong to a user and are stored in a database as belonging to a specific user Id.

Using endpoints to implement a simple API it looks like something like this prior to using the new minimal APIs:

```csharp
builder.MapGet("/api/items", async (context) =>
{
    var userId = context.User.FindFirst(ClaimTypes.NameIdentifier).Value;
    var service = context.RequestServices.GetRequiredService<ITodoService>();

    var model = await service.GetListAsync(userId, context.RequestAborted);

    await context.Response.WriteAsJsonAsync(model);
}).RequireAuthorization();
```

Converting this over to use the new minimal APIs with a daily build of ASP.NET Core 6 preview 6 gives me something like this:

```csharp
builder.MapGet("/api/items", async (ITodoService service, HttpContext context, CancellationToken cancellationToken) =>
{
    var userId = context.User.FindFirst(ClaimTypes.NameIdentifier)!.Value;
    return await service.GetListAsync(userId, cancellationToken);
}).RequireAuthorization();
```

While an improvement over the endpoint version, the `HttpContext` parameter is retained due to the need to access the user to access the claims, which in some ways seems a shame for this example because all other knowledge of the `HttpContext` has been abstracted away by the minimal APIs.

I could remove it and push the user down into the service and get it via `IHttpContextAccessor` or I could register `ClaimsPrincipal` as a scoped service to have it bound as a parameter, but they both come with the trade off that adding HttpContext accessor to DI entails.

```csharp
builder.Services.AddHttpContextAccessor();
builder.Services.AddScoped<ClaimsPrincipal>((p) =>
{
    var context = p.GetRequiredService<IHttpContextAccessor>();
    return context.HttpContext.User;
});
```

Given that `CancellationToken` has special casing in [`RequestDelegateFactory`](https://github.com/dotnet/aspnetcore/blob/dbf84eaa5a8f79947647ad64a785d39e7cd23afe/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L224-L227) and is supported via the custom [`CancellationTokenModelBinder`](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/ModelBinding/Binders/CancellationTokenModelBinder.cs) class for MVC, that made me wonder whether it would be worth doing a similar thing for `ClaimsPrincipal` in minimal actions (and MVC?) that would automatically bind `ClaimsPrincipal` to `HttpContext.User`?

Either solution changes the API endpoint to something like this:

```csharp
builder.MapGet("/api/items", async (ITodoService service, ClaimsPrincipal user, CancellationToken cancellationToken) =>
{
    var userId = user.FindFirst(ClaimTypes.NameIdentifier).Value;
    return await service.GetListAsync(userId, cancellationToken);
}).RequireAuthorization();
```

### Describe the solution you'd like

Update `RequestDelegateFactory` to support automatic binding of `ClaimsPrincipal` to `HttpContext.User`, which _seems_ like would need just two small changes.

First adding an expression to get the user from the `HttpContext`:

```csharp
private static readonly MemberExpression UserExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.User));
```

Secondly updating the [`CreateArgument(ParameterInfo, FactoryContext)`](https://github.com/dotnet/aspnetcore/blob/dbf84eaa5a8f79947647ad64a785d39e7cd23afe/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L191) method to handle a parameter of type `ClaimsPrincipal`:

```csharp
else if (parameter.ParameterType == typeof(ClaimsPrincipal))
{
    return UserExpr;
}
```

If such support was added and parity with MVC was also desired, then also adding a corresponding model binder like this which is pretty much the same implementation as the one for `CancellationToken`. In the case of MVC I guess the potential need is less than with minimal actions due to the [`ControllerBase.User`](https://github.com/dotnet/aspnetcore/blob/dbf84eaa5a8f79947647ad64a785d39e7cd23afe/src/Mvc/Mvc.Core/src/ControllerBase.cs#L222) property being readily accessible. I only thought of this in the first place because at first I thought the model binding infrastructure came into play to bind the minimal API parameters until I looked at the `RequestDelegateFactory` implementation.

```csharp
public class ClaimsPrincipalModelBinder : IModelBinder
{
    public Task BindModelAsync(ModelBindingContext bindingContext)
    {
        if (bindingContext == null)
        {
            throw new ArgumentNullException(nameof(bindingContext));
        }

        var model = (object)bindingContext.HttpContext.User;
        bindingContext.ValidationState.Add(model, new ValidationStateEntry() { SuppressValidation = true });
        bindingContext.Result = ModelBindingResult.Success(model);

        return Task.CompletedTask;
    }
}

public class ClaimsPrincipalModelBinderProvider : IModelBinderProvider
{
    private readonly ClaimsPrincipalModelBinder _modelBinder = new();

    public IModelBinder? GetBinder(ModelBinderProviderContext context)
    {
        if (context == null)
        {
            throw new ArgumentNullException(nameof(context));
        }

        if (context.Metadata.ModelType == typeof(ClaimsPrincipal))
        {
            return _modelBinder;
        }

        return null;
    }
}
```

If support for this was deemed to be worth adding, I'd be happy to submit a PR to do so.

### PR DESCRIPTION
**Pull request by martincostello**
**PR Title**

Support Request, Response and User for minimal actions

**PR Description**

Adds support to minimal actions for parameters of type `HttpRequest`, `HttpResponse`, and `ClaimsPrincipal` to be bound to the values of the `Request`, `Response` and `User` properties of the `HttpContext` respectively.

Also cleans up some typos and IDE suggestions in the `RequestDelegateFactory` tests.

Fixes #33870

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -221,6 +223,18 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
...
+                return HttpRequestExpr;
+            }
+            else if (parameter.ParameterType == typeof(HttpResponse))
+            {
+                return HttpResponseExpr;
+            }
+            else if (parameter.ParameterType == typeof(ClaimsPrincipal))
+            {
+                return UserExpr;
+            }
```
Comment (martincostello) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:237]: I wasn't sure what the best sequencing might be for the conditions here for if there was an optimal order from a perf perspective, so this order is kinda arbitrary.

Comment (davidfowl) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:237]: It doesn't *really* matter since this happens once at startup.

**Review by davidfowl — APPROVED** [review]
LGTM, but will give @halter73 a chance to review.

**Comment by davidfowl** [comment]
@martincostello one thing we missed is the api explorer support https://github.com/dotnet/aspnetcore/blob/26fba6585ccf238958f5d2f28715eb53b06a8ae1/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L169. Right now these will show up as coming from the body isn't correct.

**Comment by martincostello** [comment]
Good spot. Do you want me to push up a PR to amend that?

**Comment by davidfowl** [comment]
> Good spot. Do you want me to push up a PR to amend that?

Yep!

**Pull request merged**